README.md
#### Download the latest release: Somatypus 1.3  ( [zip](../../archive/v1.3.zip) | [tar.gz](../../archive/v1.3.tar.gz) ).

---


Somatypus
=========

### A Platypus-based variant calling pipeline for cancer data

__Adrian Baez-Ortega  
Transmissible Cancer Group, University of Cambridge__

Somatypus is an open-source pipeline that makes use of the powerful variant caller Platypus for identifying germline and somatic SNVs and indels in sequencing data coming from a set of unpaired samples. It has been designed to offer great sensitivity and specificity, even for low-frequency somatic mutations found in cancer genomes. Although this pipeline can be applied to any cancer sequence data, it is particularly useful for the processing of data obtained from many tumour samples that are closely related to each other and lack matched normal samples, such as transmissible cancer or metastasis data.

Somatypus has been tested on Ubuntu (14.04.4) systems, and it should work well on any Linux distribution. It has not been tested on Mac systems, although it may work after some minor code modifications.


#### Somatypus does...

* Call single nucleotide variants (SNVs) and short insertions/deletions (indels) from any number of sequence alignment files (subject to memory constraints), at a probably unmatched speed.

* Filter low-quality or ambiguous variants, while keeping those variants occurring with low frequency (in a few samples) or aberrant copy number (not fitting a diploid model).

* Allow the use of additional calling options for Platypus, as long as they do not override the ones defined in the pipeline.

* Run seamlessly from a single command.

* Resume execution after an unexpected interruption.

#### Somatypus does not...

* Call long indels, structural variants or copy number changes.

* Manage samples in pairs, or distinguish between “tumour” and “host/normal” samples (or between germline and somatic variants). Identification of somatic variants must be performed downstream.

* Take into account cross-sample contamination. Identification of variants caused by contamination must be performed downstream.

* Allow extensive workflow customisation (unless the source code is altered).


---

## Installation

(The following information can be found in the INSTALL.txt file.)

Somatypus has been designed as a set of Bash and Python scripts that do not require a real 
installation, but simply adding the software directory to the PATH environment variable, 
thus making set-up straightforward. However, it has some minimal software dependencies:

* __Python 2.6 or later__ (not tested on Python 3.X), including development libraries.

* __Platypus__ (http://www.well.ox.ac.uk/platypus), which in turn requires __htslib__ (http://www.htslib.org), which in turn requires __zlib__ (http://zlib.net).

* __VCFtools__ (https://vcftools.github.io), for using the vcf-sort command.

* The __tabix__ package (http://sourceforge.net/projects/samtools/files/tabix), for using the tabix and bgzip commands. It should come together with htslib.


For the sake of tidiness, it is advisable to create a folder called "somatypus" in the
location where you usually install software, and install all the dependencies and the
Somatypus pipeline itself (the somatypus-x.x folder) into it.


The following instructions assume that you have administrator privileges in the system 
you are using (i.e. that you can do "sudo"); otherwise, you should contact you system's
administrator. __Although it should be obvious, please note that you need to replace
things like "path/to/" and "/FULL/PATH/TO/" with the actual path to the relevant folder.
"/FULL/PATH/TO/" indicates that the absolute path (beginning at "/", e.g. /home/user/...)
must be used.__


The recommended installation order is:


 1. __Python ≥2.6__
 
    Python is normally installed by default in most UNIX systems. However, in order to run Platypus, you need to install the Python development libraries.  
    On Debian or Ubuntu Linux, you can install them with:
        
        sudo apt-get install python-dev
    
    On RPM-based Linux distributions, you can install them with:

        sudo yum install python-devel


 2. __zlib (development files)__
 
    If you have trouble installing htslib (below), you probably need to install this first.  
    On Debian or Ubuntu Linux, you can install the corresponding package with:

        sudo apt-get install zlib1g-dev
    
    On RPM-based Linux distributions, you can install it with:

        sudo yum install zlib-devel


 3. __htslib__
 
    This is necessary in order to run Platypus.  
    Latest version as of April 2016: https://github.com/samtools/htslib/releases/download/1.3/htslib-1.3.tar.bz2  
    Once downloaded and uncompressed, you can install it with:

        cd path/to/htslib-x.x
        ./configure
        make
        sudo make install

    After this, you need to add the htslib directory to your LD_LIBRARY_PATH environment variable. You can do this either by editing your ~/.bashrc file with a text editor (e.g. nano) and adding the following line:

        export LD_LIBRARY_PATH=/FULL/PATH/TO/htslib-x.x:$LD_LIBRARY_PATH

    Or just by appending the relevant line to the file with:

        cd path/to/htslib-x.x
        echo "export LD_LIBRARY_PATH=$PWD:\$LD_LIBRARY_PATH" >> ~/.bashrc
   
    Either way, it is then necessary to source the .bashrc file for the changes to be applied:

        source ~/.bashrc


 4. __Platypus__
 
    This is the best and the worst part of the pipeline.  
    Latest version: http://www.well.ox.ac.uk/software-download-registration  
    Once downloaded and uncompressed, you can install it with:

        cd path/to/Platypus_x.x.x
        ./buildPlatypus.sh

    If installation is successful, you will see the message "Finished building Platypus". After this, you need to make a symbolic link to the Platypus executable file in the somatypus-x.x/src folder (which will be then added to the PATH variable, see below).

        cd path/to/Platypus_x.x.x
        ln -s $PWD/Platypus.py path/to/somatypus-x.x/src/


 5. __VCFtools__
 
    On Debian or Ubuntu Linux, you can install the corresponding package with:

        sudo apt-get install vcftools
    
    On RPM-based Linux distributions, you can install it with:

        sudo yum install vcftools
    

 6. __Somatypus__
 
    The last step is adding the Somatypus directory to your PATH environment variable, so that the somatypus command can be called from the command line. You can do this either by editing your ~/.bashrc file with a text editor (e.g. nano) and adding the line:

        export PATH=/FULL/PATH/TO/somatypus-x.x/src:/FULL/PATH/TO/somatypus-x.x/utils:$PATH

    Or just by appending the relevant line to the file with:

        cd path/to/somatypus-x.x
        echo "export PATH=$PWD/src:$PWD/utils:\$PATH" >> ~/.bashrc
   
    Either way, it is then necessary to source the .bashrc file for the changes to be applied:
    
        source ~/.bashrc


Once all the dependencies and the pipeline have been installed, you should be able to run all of the following commands (which show the usage information of each tool):

    Platypus.py callVariants -h
    tabix
    bgzip
    vcf-sort -h
    somatypus

And now you can have fun.


---

## Running Somatypus

When the pipeline is run via the `somatypus` command with no options (or with `-h`), it displays the usage information.

    | SOMATYPUS
    | A Platypus-based variant calling pipeline for cancer data
    | Version x.x
    |
    | Required input:
    |    -i  Absolute path to folder containing the input BAM files (accompanied by BAI indices).
    |    -g  Absolute path to reference genome FASTA file (accompanied by FAI index).
    |    -o  Absolute path to the output folder (it will be created if needed).
    |
    | Optional input:
    |    -r  Absolute path to file of regions to use, one per line in CHR:START-END format.
    |    -c  Number of CPUs (processes) for Platypus *(should not exceed 8 due to a bug)*.
    |    -p  Additional options for Platypus, within quotes and separated by spaces.
    |
    | Options:
    |    -w  Use windows around the variants as regions during genotyping.
    |    -h  Print this usage information and exit.
    |    -v  Print version and exit.
    |
    | Usage:
    |    somatypus -i /path/to/bams_dir -o /path/to/out_dir -g /path/to/genome.fna -r /path/to/regions.txt -c <1-8> -p "--option=VAL --option=VAL"

It is advisable that all the input paths be absolute, rather than relative. 

An optional regions file (`-r` option) allows the user to define a set of genomic regions (e.g. exons) wherein to perform the calling. The regions file must be a text file containing one region per line, in CHR:START-END format (e.g. 1:1028676-1028844. The chromosome names must match those in the FASTA and BAM files).

The `-w` option enables the use of windows of +/-200 bp around each variant as regions for genotyping. This may increase the efficiency of the genotyping process, especially in cases where no regions file is used. However, since it may cause some variants to be missed, this option should not be used unless necessary.

The number of CPUs is also optional (default is 1) but, if specified, must be at least 1, and should not exceed 8 (or even less, depending on the amount of data), due to an inveterate Platypus bug that can cause an extremely excessive memory allocation attempt (see the [full documentation](docs/Somatypus%20Documentation.pdf)).

Finally, additional calling options can be passed to Platypus via the `-p` option. The entire additional options string must be quoted, and options must be separated by spaces. However, those options already specified in the pipeline cannot be included, namely: `--logFileName`, `--refFile`, `--bamFiles`, `--regions`, `--minPosterior`, `--minReads`, `--minFlank`, `--trimReadFlank`, `--source`, `--getVariantsFromBAMs`, `--nCPU`, or `--output` (or `-o`). (For obvious reasons, they should also not include `--help` or `-h`.)

A list of all the Platypus options can be consulted via: `Platypus.py callVariants -h`.

The full log of the pipeline execution will be stored in a file named SOMATYPUS_<*date+time*>.log, in the logs subfolder of the output directory, together with the logs of most of the steps. The log files and the temporary folders containing intermediate files will be numbered according to the number of the steps that interact with them.


### For more information on the workflow and output files, please read the full documentation in [docs/Somatypus Documentation.pdf](docs/Somatypus%20Documentation.pdf).


---

## Licence

Copyright © 2016–2017 Transmissible Cancer Group, University of Cambridge  
Author: Adrian Baez-Ortega ([ORCID 0000-0002-9201-4420] (http://orcid.org/0000-0002-9201-4420); ab2324@cam.ac.uk)

Somatypus is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation; either version 3 of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with this program. If not, see http://www.gnu.org/licenses.


docs/Changelog.txt
# Change Log

## [1.3] - 2017-02-03

### Added
- Command-line option -w, which enables the use of windows of +/-200 bp around each
  variant as regions for genotyping. This may increase the efficiency of the 
  genotyping process, especially in cases where no regions file is used. However,
  since it may cause some variants to be missed, this option should not be used
  unless necessary.

### Changed
- Quality filtering for called/genotyped indels has been modified, so that 
  filtering based on Platypus quality flags is done in the same manner as for SNVs.
- Workflow updated to allow cases where no indels are present in the data.
- Workflow updated to allow cases where no indel-excluded SNVs are present in the data.
- Improved code for merging of indel regions before genotyping (by Kevin Gori).
- Fixed bug in median VAF calculation during the filtering of indel-flagged SNVs.
- Minor change in the naming of intermediate genotyped VCF files.
- Documentation updated.


## [1.2.1] - 2016-09-21

### Changed

- 'SNP' changed to 'SNV' in all the intermediate file names, and in script names.
- Other minor changes in scripts.
- Documentation updated.


## [1.2] - 2016-05-14

### Added
- Somatypus_MergeIndelRegions.py script, which merges overlapping regions from the 
  regions file used for re-genotyping indels that are missing after genotyping.

### Changed
- Indels are now genotyped separately in an additional pipeline step (#12). This
  solves a problem that caused a small proportion of variant calls to be duplicated.
- Indices of steps after #11 have been updated after the inclusion of indel 
  genotyping.
- Intermediate region files called 'SNPregions_...' renamed to 'varRegions_...'.
- Somatypus_VafFilter.py script no longer performs filtering of indels according to 
  quality flags, or separation of SNPs and indels.
- Somatypus_ExtractExons.py script renamed to Somatypus_ExtractRegions.py.
- Somatypus_SNPs_final.vcf output file renamed to Somatypus_SNVs_final.vcf.
- Slightly different naming of genotyping log files.
- Output files are now sorted by chromosome in natural order: 
  1,2,...,10,11,...,20,21,...,X instead of 1,10,11,...,2,20,21,...,X
- Documentation updated.


## [1.1] - 2016-04-15

### Added
- Command-line option -p for specifying additional Platypus arguments.
- Subsequent Somatypus executions with the same output folder (to resume execution) 
  generate different global log files, labeled with date and time.
- Individual variant calling generates a separate log file for each sample.

### Changed
- Temporary folders and log files naming convention, so that the numbers in
  the folder/file names match those of the step(s) involved.
- Detailed installation steps in INSTALL.txt.
- Command-line usage information.
- Other minor changes to the somatypus script.
- Documentation updated.

### Removed
- Installation script (more dangerous than useful).


install_somatypus.sh
# INSTALLING SOMATYPUS AND DEPENDENCIES
# Adrian Baez-Ortega
# 14/04/2016


# This script installs Somatypus and all its dependencies (in a personal computer).
#
# Before running this script, you should:
#
#   1. FIRST OF ALL, make sure you have the correct Python dependencies by running (Ubuntu):
#        sudo apt-get update
#        sudo apt-get install python-dev libxml2-dev libxslt-dev
#
#   2. Unpack the somatypus-1.0.tgz in the directory in which you want to install the 
#      software (e.g. /home/your_user/software).
#
# Then, you need to run this install script (replace the example path with your own):
#
#   /PATH/TO/somatypus-1.0/install_somatypus.sh
#
# You will be asked for your user password at some point of the process; this is because 
# you need administratorprivileges in order to install some dependencies (more concretely, 
# for running the 'make install' commands). If you do not have administrator privileges in 
# your computer, please follow alternative instructions in the INSTALL.txt file.



# Go to the installation directory
INIDIR=$PWD
cd "$( dirname "${BASH_SOURCE[0]}" )"



# Installation of software dependencies
LIBSTR="\$LD_LIBRARY_PATH"
PATHSTR="\$PATH"

# (If you can run "Platypus.py callVariants -h", then go to step 2) 
# 1. Download and install Platypus
hash Platypus.py 2>/dev/null || {   

    echo -e "\nInstalling Platypus\n"
    wget http://www.well.ox.ac.uk/bioinformatics/Software/Platypus-latest.tgz
    tar -zxvf Platypus-latest.tgz
    rm Platypus-latest.tgz
    cd Platypus_*


    # Download and install htslib (Platypus dependency)
    wget https://github.com/samtools/htslib/releases/download/1.3/htslib-1.3.tar.bz2
    tar xvjf htslib-1.3.tar.bz2
    rm htslib-1.3.tar.bz2
    mv htslib-1.3 htslib
    cd htslib
    ./configure
    make
    sudo make install    # needs admin privileges
    LIBSTR="$PWD:$LIBSTR"
    # To try htslib: type "tabix" and "bgzip"


    # Install Platypus
    cd ..
    ./buildPlatypus.sh
    PATHSTR="$PWD:$PATHSTR"
    export PATH=$PWD:$PATH
    cd ..

}
# To try Platypus: type "Platypus.py callVariants -h"
hash Platypus.py 2>/dev/null || { echo -e "\nThere was a problem installing Platypus. If the lines above do not provide further information, please run this script line-by-line to find the problem.\n" >&2; exit 1; }



# (If you can run "tabix" and "bgzip", then go to step 3)
# 2. Download and install tabix (htslib)
hash tabix 2>/dev/null || { 

    echo -e "\nInstalling tabix (htslib)\n"
    wget https://github.com/samtools/htslib/releases/download/1.3/htslib-1.3.tar.bz2
    tar xvjf htslib-1.3.tar.bz2
    rm htslib-1.3.tar.bz2
    mv htslib-1.3 htslib
    cd htslib
    ./configure
    make
    sudo make install    # needs admin privileges
    cd ..

}    
# To try htslib: type "tabix" and "bgzip"
hash tabix 2>/dev/null || { echo -e "\nThere was a problem installing tabix. If the lines above do not provide further information, please run this script line-by-line to find the problem.\n" >&2; exit 1; }
hash bgzip 2>/dev/null || { echo -e "\nThere was a problem installing bgzip. If the lines above do not provide further information, please run this script line-by-line to find the problem.\n" >&2; exit 1; }



# (If you can run "vcf-sort -h", then go to step 4)
# 3. Install VCFtools
hash vcf-sort 2>/dev/null || {

    echo -e "\nInstalling VCFtools\n"
    wget https://github.com/vcftools/vcftools/releases/download/v0.1.14/vcftools-0.1.14.tar.gz
    tar -zxvf vcftools-0.1.14.tar.gz 
    rm vcftools-0.1.14.tar.gz
    cd vcftools-0.1.14
    ./configure
    make
    sudo make install    # needs admin privileges
    cd ..

}
# To try VCFtools: type "vcf-sort -h"
hash vcf-sort 2>/dev/null || { echo -e "\nThere was a problem installing VCFtools. If the lines above do not provide further information, please run this script line-by-line to find the problem.\n" >&2; exit 1; }


# (If you can run "Somatypus_SplitMA-MNPs.py", then skip this step)
# 4. Add Somatypus to the PATH
hash somatypus 2>/dev/null || {
    
    echo -e "\nInstalling Somatypus\n"
    PATHSTR="$PWD/src:$PWD/utils:$PATHSTR"
    export PATH=$PWD/src:$PWD/utils:$PATH
    
}
# To try Somatypus: "somatypus"
hash somatypus 2>/dev/null || { echo -e "\nThere was a problem adding the Somatypus directories to the PATH. If the lines above do not provide further information, please run this script line-by-line to find the problem.\n" >&2; exit 1; }    

echo "export LD_LIBRARY_PATH=$LIBSTR" >> ~/.bashrc
echo "export PATH=$PATHSTR" >> ~/.bashrc

cd $INIDIR
echo -e "\nInstallation completed!\nNow, type:\n\nsource ~/.bashrc\nsomatypus\n"

